File: trivially/src/main/java/trivially/ep/m6/Neg.java
package trivially.ep.m6;

import trivially.ep.m6.finalized.Lit;

public interface Neg extends Exp, trivially.ep.m5.Neg {

    Exp getInner();

    default Boolean equals(Exp other) {
        return this.astree().equals(other.astree());
    }

    default Exp simplify() {
        if (this.getInner().eval().equals(0.0)) {
            return new Lit(0.0);
        } else {
            return new trivially.ep.m6.finalized.Neg(this.getInner().simplify());
        }
    }

    default Boolean isNeg(Exp inner) {
        return inner.eql(getInner());
    }

    default Boolean eql(Exp that) { return that.isNeg(getInner()); }

    default void truncate (int level) {
        if (level > 1) {
            getInner().truncate(level-1);
        } else {
            setInner(new trivially.ep.m5.finalized.Lit(getInner().eval()));
        }
    }
}

File: coco_csharp/CoCo/EP/M6/Neg.cs
namespace EP.M6 {
    public interface Neg<FT> : Exp<FT>, EP.M5.Neg<FT> {
        bool Exp<FT>.equals(EP.Exp<FT> other) {
            return convert(this).astree().Equals(convert(other).astree());
        }

        bool Exp<FT>.isNeg(EP.Exp<FT> inner) {
            return convert(inner).eql(getInner());
        }

        bool Exp<FT>.eql(EP.Exp<FT> that) {
            return convert(that).isNeg(getInner());
        }
    }
}